lib/Dialect/TensorExt/Transforms/ImplementRotateAndReduce.cpp

```diff
@@ -0,0 +1,156 @@
+  }
+
+  IRRewriter rewriter(op.getContext());
+  TypedValue<RankedTensorType> input = op.getTensor();
+  TypedValue<RankedTensorType> plaintexts = op.getPlaintexts();
+  unsigned steps = op.getSteps().getZExtValue();
+  unsigned period = op.getPeriod().getZExtValue();
+
+  StringRef mulOpName = isa<IntegerType>(input.getType().getElementType())
+                            ? "arith.muli"
+                            : "arith.mulf";
+  StringRef addOpName = isa<IntegerType>(input.getType().getElementType())
+                            ? "arith.addi"
+                            : "arith.addf";
+
+  // Use a value of sqrt(n) as the baby step / giant step size.
+  auto babySteps = static_cast<int64_t>(std::floor(std::sqrt(steps)));
+  unsigned giantSteps = steps / babySteps;
+  if (giantSteps * babySteps != steps) {
+    return op.emitOpError()
+           << "requires steps to be a multiple of sqrt(steps), but found "
+              "steps="
+           << steps << " and babySteps=" << babySteps;
+  }
+  LLVM_DEBUG(llvm::dbgs()
+             << "Using baby-step / giant-step decomposition of sum of size "
+             << steps << " with babySteps= " << babySteps
+             << " and giantSteps= " << giantSteps << "\n");
+
+  // Compute sqrt(n) ciphertext rotations of the input as baby-steps.
+  rewriter.setInsertionPointAfter(op);
+  SmallVector<Value> babyStepVals;
+  babyStepVals.push_back(input);
+  for (int64_t i = 1; i < babySteps; ++i) {
+    babyStepVals.push_back(rewriter
+                               .create<tensor_ext::RotateOp>(
+                                   op->getLoc(), input,
+                                   rewriter.create<arith::ConstantIndexOp>(
+                                       op->getLoc(), period * i))
+                               .getResult());
+  }
+
+  unsigned plaintextSize = plaintexts.getType().getRank();
+  SmallVector<OpFoldResult> offsets(plaintextSize, rewriter.getIndexAttr(0));
+  SmallVector<OpFoldResult> sliceSizes;
+  sliceSizes.reserve(plaintextSize);
+  sliceSizes.push_back(rewriter.getIndexAttr(1));
+  for (int64_t i = 1; i < plaintextSize; ++i) {
+    sliceSizes.push_back(
+        rewriter.getIndexAttr(plaintexts.getType().getDimSize(i)));
+  }
+  SmallVector<OpFoldResult> unitStrides(plaintextSize,
+                                        rewriter.getIndexAttr(1));
+
+  // Compute the inner baby step sums.
+  Value result;
+  for (unsigned k = 0; k < giantSteps; ++k) {
+    Value innerSum;
+    for (unsigned j = 0; j < babySteps; ++j) {
+      offsets[0] = rewriter.getIndexAttr(j + k * babySteps);
+      Value rotatedPlaintext = rewriter.create<tensor_ext::RotateOp>(
+          op->getLoc(),
+          rewriter.create<tensor::ExtractSliceOp>(op->getLoc(), input.getType(),
+                                                  plaintexts, offsets,
+                                                  sliceSizes, unitStrides),
+          rewriter.create<arith::ConstantIndexOp>(op->getLoc(),
```

Comment:
nit: This index op can be factored out of the loop; slightly more efficient to not have to cse them later.